src/vcSHP.cpp

```diff
@@ -0,0 +1,740 @@
+      readPosition = UnpackDouble(readPosition, &record.data.multiPointM.MRange.y, buffer);
+
+      record.data.multiPointM.MValue = udAllocType(double, record.data.multiPointM.pointsCount, udAF_Zero);
+      UD_ERROR_NULL(record.data.multiPointM.MValue, udR_MemoryAllocationFailure);
+      for (int i = 0; i < record.data.multiPointM.pointsCount; i++)
+        readPosition = UnpackDouble(readPosition, &record.data.multiPointM.MValue[i], buffer);
+    }
+  }
+  break;
+  default:
+  break;
+  }
+
+  pSHP->shpRecords.PushBack(record);
+  offset += readPosition;
+
+epilogue:
+  udFree(buffer);
+
+  if (result != udR_Success)
+    vcSHP_ReleaseRecord(&record);
+
+  return result;
+}
+
+udResult vcSHP_LoadShpFile(vcSHP *pSHP, const char *pFilename)
+{
+  udResult result;
+
+  udFile *pFile = nullptr;
+  uint8_t *headerBuffer = nullptr;
+  uint32_t offset = 0;
+
+  vcSHP_Header &header = pSHP->shpHeader;
+  UD_ERROR_CHECK(udFile_Open(&pFile, pFilename, udFOF_Read));
+
+  headerBuffer = udAllocType(uint8_t, 100, udAF_Zero);
+  UD_ERROR_NULL(headerBuffer, udR_MemoryAllocationFailure);
+  UD_ERROR_CHECK(udFile_Read(pFile, headerBuffer, 100 * sizeof(uint8_t)));
+  // 00 00 27 0a
+  UD_ERROR_IF((headerBuffer[0] != 0 || headerBuffer[1] != 0 || headerBuffer[2] != 0x27 || headerBuffer[3] != 0x0a), udR_ParseError);
+
+  // the file stores the length in shorts
+  header.fileLength = (headerBuffer[24] << 24) | (headerBuffer[25] << 16) | (headerBuffer[26] << 8) | headerBuffer[27];
+  header.fileLength = 2 * udMin(header.fileLength, UINT32_MAX / 2);
+
+  header.version = (headerBuffer[31] << 24) | (headerBuffer[30] << 16) | (headerBuffer[29] << 8) | headerBuffer[28];
+  header.shapeType = (vcSHPType)((int32_t)((headerBuffer[35] << 24) | (headerBuffer[34] << 16) | (headerBuffer[33] << 8) | headerBuffer[32]));
+
+  offset = 36;
+  offset = UnpackDouble(offset, &header.MBRmin.x, headerBuffer);
+  offset = UnpackDouble(offset, &header.MBRmin.y, headerBuffer);
+  offset = UnpackDouble(offset, &header.MBRmax.x, headerBuffer);
+  offset = UnpackDouble(offset, &header.MBRmax.y, headerBuffer);
+  offset = UnpackDouble(offset, &header.ZRange.x, headerBuffer);
+  offset = UnpackDouble(offset, &header.ZRange.y, headerBuffer);
+  offset = UnpackDouble(offset, &header.MRange.x, headerBuffer);
+  offset = UnpackDouble(offset, &header.MRange.y, headerBuffer);
+
+  UD_ERROR_IF(offset != 100, udR_ParseError);
+  udFree(headerBuffer);
+
+  UD_ERROR_CHECK(pSHP->shpRecords.Init(32));
+  while (offset < header.fileLength)
+  {
+    result = vcSHP_LoadShpRecord(pSHP, pFile, offset, header.shapeType);
+    UD_ERROR_CHECK(result);
+  }
+
+epilogue:
+  udFile_Close(&pFile);
+
+  if (headerBuffer)
+    udFree(headerBuffer);
+
+  return result;
+}
+
+udResult vcSHP_LoadFileGroup(vcSHP *pSHP, const udFilename &fileName)
+{
+  if (pSHP == nullptr || !fileName.HasFilename())
+    return udR_InvalidParameter_;
+
+  udFilename dbfName = fileName;
```

Comment:
I had intended for you to only change the internals of this function rather than passing a `udFilename` in, just using one internally. You don't need multiple `udFilename` objects, you can just modify the one file. Create the `udFilename`, load the shapefile, call `SetExtension`, load the DBF, call `SetExtension`, load the PRJ. Does `udFilename` work with this assignment operator, there isn't an explicit one but maybe it's converting to `const char *` and using that constructor.

Reply:
fixed.